### DISCUSSION TITLE
How to implement a second row type for sub rows?

### DISCUSSION THREAD

**Discussion by varwasabi** [TanStack/table#4468] [Q&A]
I'm using the expanded rows functionality but on the expanded rows i have a differing data set. 

Using the same cell method in the columnDefs sticks to the first type but i need a way for it to reference the subRow type. i.e. when in the cell if row.depth is 1 then everything on row would need to reference the second type otherwise it would work as is out of the box.

**Comment by Doesntmeananything** [comment]
I have a similar use case. In my table, I'd like to have parent rows to be a simple label with a button to toggle expanded state. Ideally, these would be a different type compared to children rows.

For now, I work around this by making the parent rows the same type as the children rows, however that leads to a number of type issues as the table grows in complexity.

Essentially, I'm looking for a way to define the inverse of sub-components (as shown [here](https://tanstack.com/table/v8/docs/examples/react/sub-components)), where the parent row is a separate type that is still defined within the table data model.

**Comment by raphaelboukara** [comment]
Hi :)
I have the exact same use case where the type of the sub rows is not the same like the parents. 

ex:
```ts
type data = { id: string, firstname: string, comments: { id: string, content: string }[] }[];
```

My table looks like this:
```
______________________________________________________________
|                                   |     comment 1
|  user1 (rowSpan 3)                |     comment 2
|                                   |     comment 3
______________________________________________________________
|                                   |     comment 4
|  user2 (rowSpan 3)                |     comment 5
|                                   |     comment 6
______________________________________________________________
```

Maybe it does not make sense to use sub rows feature to handle this use case?

**Reply by BrendanC23** [reply]
I don't think there's a great way to generate a table with TanStack Table when the subrow data has a different data type than the main data. One alternative is to un-nest the comments so that each row of the table data is a separate comment. You can then group by the `firstname` property.

```tsx
type Data = {
    id: string;
    firstname: string;
    comments: {
        id: string;
        content: string;
    }[]
};

const data = Array(10)
    .fill(undefined)
    .map<Data>((_, i) => ({
        id: `user_${i}_id`,
        firstname: `User ${i} Name`,
        comments: Array(i)
            .fill(undefined)
            .map((_comment, j) => ({ id: `user_${i}_comment_${j}`, content: `User ${i} Comment ${j}` })),
    }));

const flatData = data.flatMap((item) =>
    item.comments.map((comment) => ({
        id: item.id,
        firstname: item.firstname,
        comment,
    })),
);

// { id: string; firstname: string; comment: { id: string, content: string } }[];
type FlatData = (typeof flatData)[number];

const columnHelper = createColumnHelper<FlatData>();
const columns = [
    columnHelper.accessor("firstname", {
        header: "First Name",
    }),
    columnHelper.accessor("comment.content", {
        header: "Comment",
    }),
];

export function App() {
    return (
        <ReactTable
            data={flatData}
            columns={columns}
            grouping={["firstname"]}
        />
    );
}
```

<img width="1004" height="731" alt="image" src="https://github.com/user-attachments/assets/d9bdb2bb-3070-4706-ad3e-cb1e9623b16d" />


This works fairly well for simple tables. It would be great if there was a better way that didn't require manually reformatting the data. This approach is cumbersome if there are multiple layers of nesting, like the following data structure. In this case, the data would need to be flattened at multiple levels and you would need to be able to group by multiple columns.


```tsx
type Data = {
    users: User[];
}

type User = {
    firstName: string;
    lastName: string;
    posts: Post[];
}

type Post = {
    title: string;
    subtitle: string;
    body: string;
    comments: Comment[];
}

type Comment = {
    date: string;
    comment: string;
}
```
